### DISCUSSION TITLE
Suggestion: Parsing incomplete and buggy IFC files

### DISCUSSION THREAD

**Discussion by chetan-set** [IfcOpenShell/IfcOpenShell#3257] [Ideas]
Hello everyone,

I have been dealing with a strange IFC file that renders fine on Solibri but crashes on IFCConvert. After I spent some time inspecting the file, I realised that the file itself was buggy. In a nutshell, it was using `IFCRELASSOCIATESMATERIAL` without actually specifying the material.

#### Here is my proposal:

Can we make IFCConvert more flexible and do its best to render incomplete/buggy files?

#### Current behaviour:

As soon as a mandatory parameter is not found, an exception is thrown and IFCConvert terminates.

#### Worry owner:

I propose to be the person to do the actual work. I will modify the ifc parser to handle buggy/incomplete ifc files.

Please give me your thoughts.

#### Additional information

Here is some more information the crash I am experiencing:

I have truncated the reduced the IFC file to just 81 lines which I am reproducing at the end of this message.

To fix the file, two lines are needed. A `IFCMaterial` definition:

`#76= IFCMATERIAL('Solid');`

And an update to the  `IFCRELASSOCIATESMATERIAL`  as such:

`#52=IFCRELASSOCIATESMATERIAL('1kN1AIT199XQkai3RCRaow',#1,'IfcRelAssociatesMaterial',$,(#55),#76);`

The IFC file:

```
ISO-10303-21;
HEADER;
FILE_DESCRIPTION(('model test'), '2;1');
FILE_NAME('','2000-01-01T00:00:00',(),(), '', '', '');
FILE_SCHEMA(('IFC2X3'));
ENDSEC;
DATA;
#1=IFCOWNERHISTORY(#2,#5,$,$,$,$,$,1);
#2=IFCPERSONANDORGANIZATION(#3,#4,$);
#3=IFCPERSON('XUSER','XUSER','XUSER',$,$,$,$,$);
#4=IFCORGANIZATION('XXX','XXX',$,$,$);
#5=IFCAPPLICATION(#4,'V1.0','X','X');
#6=IFCMATERIALLAYERSET((),$);
#7=IFCPROJECT('c6cda6bffd5e4239ac8b076fdc4960a2',#1,'X',$,$,'X',$,(),#8);
#8=IFCUNITASSIGNMENT((#9));
#9=IFCSIUNIT(*,.LENGTHUNIT.,$,.METRE.);
#10=IFCSITE('412b75f359b648ce92bd434d306b1244',#1,'X site',$,$,#11,#16,$,.ELEMENT.,$,$,$,$,$);
#11=IFCLOCALPLACEMENT($,#12);
#12=IFCAXIS2PLACEMENT3D(#13,$,$);
#13=IFCCARTESIANPOINT((0.0,0.0,0.0));
#14=IFCDIRECTION((0.0,0.0,1.0));
#15=IFCDIRECTION((1.0,0.0,0.0));
#16=IFCPRODUCTREPRESENTATION($,$,(#17));
#17=IFCSHAPEREPRESENTATION($,'FootPrint','Curve2D',(#18));
#18=IFCPOLYLINE((#19,#20,#21,#22,#19));
#19=IFCCARTESIANPOINT((-37121.33,-8261.41));
#20=IFCCARTESIANPOINT((25662.47,-7262.41));
#21=IFCCARTESIANPOINT((16163.47,9623.037));
#22=IFCCARTESIANPOINT((-74213.33,624.037));
#23=IFCRELAGGREGATES('8829336caaad4a74a8f8ebf0e8a18416',#1,'Aggregate0',$,#7,(#10));
#24=IFCBUILDING('a44fd80e82214683bb7b584b2231a37d',#1,'X building',$,$,#25,#30,$,.ELEMENT.,$,$,$);
#25=IFCLOCALPLACEMENT($,#26);
#26=IFCAXIS2PLACEMENT3D(#27,$,$);
#27=IFCCARTESIANPOINT((0.0,0.0,0.0));
#28=IFCDIRECTION((0.0,0.0,1.0));
#29=IFCDIRECTION((1.0,0.0,0.0));
#30=IFCPRODUCTREPRESENTATION($,$,(#31));
#31=IFCSHAPEREPRESENTATION($,'FootPrint','Curve2D',(#32));
#32=IFCPOLYLINE((#33,#34,#35,#36,#33));
#33=IFCCARTESIANPOINT((-86911.207,-4120.092));
#34=IFCCARTESIANPOINT((34822.34,-3220.08));
#35=IFCCARTESIANPOINT((34833.34,2355.18));
#36=IFCCARTESIANPOINT((-86944.27,1466.71));
#37=IFCRELAGGREGATES('43411a7f453c4310bff32e54c2e1d9eb',#1,'Aggregate0',$,#10,(#24));
#38=IFCBUILDINGSTOREY('63cc8a3756f04a208a216aaf0d06326c',#1,'floor#1',$,$,#39,#44,$,.ELEMENT.,-5.0);
#39=IFCLOCALPLACEMENT($,#40);
#40=IFCAXIS2PLACEMENT3D(#41,$,$);
#41=IFCCARTESIANPOINT((10.0,20.0,30.0));
#42=IFCDIRECTION((0.0,0.0,1.0));
#43=IFCDIRECTION((1.0,0.0,0.0));
#44=IFCPRODUCTREPRESENTATION($,$,(#45));
#45=IFCSHAPEREPRESENTATION($,'FootPrint','Curve2D',(#46));
#46=IFCPOLYLINE((#47,#48,#49,#50,#47));
#47=IFCCARTESIANPOINT((-622011.34,-451.75));
#48=IFCCARTESIANPOINT((870802.67,-872.75));
#49=IFCCARTESIANPOINT((540803.38,3493.4021532592117));
#50=IFCCARTESIANPOINT((-416014.078642035434,5493.4021532592117));
#51=IFCRELAGGREGATES('e9af8633ae44449daa8133b1fb201546$',#1,'Aggregate0',$,#24,(#38));
#52=IFCRELASSOCIATESMATERIAL('1kN1AIT199XQkai3RCRaow',#1,'IfcRelAssociatesMaterial',$,(#55),$);
#55=IFCBEAM('478ba5c0464b4fb9b63179c30dc05b27',#1,'Beam#1',$,$,#56,#61,$);
#56=IFCLOCALPLACEMENT($,#57);
#57=IFCAXIS2PLACEMENT3D(#58,$,$);
#58=IFCCARTESIANPOINT((10.0,20.0,30.0));
#59=IFCDIRECTION((0.9998234381619838,0.0,0.0187907557578117));
#60=IFCDIRECTION((0.0,1.0,0.0));
#61=IFCPRODUCTREPRESENTATION($,$,(#62));
#62=IFCSHAPEREPRESENTATION($,'Body','SweptSolid',(#63));
#63=IFCEXTRUDEDAREASOLID(#68,#64,#74,4.050715201721251);
#64=IFCAXIS2PLACEMENT3D(#65,#66,#67);
#65=IFCCARTESIANPOINT((-410161.0,50.0,3452.149109));
#66=IFCDIRECTION((0.9998234381619838,0.0,0.0187907557578117));
#67=IFCDIRECTION((0.0,1.0,0.0));
#68=IFCARBITRARYCLOSEDPROFILEDEF(.AREA.,$,#69);
#69=IFCPOLYLINE((#70,#71,#72,#73,#70));
#70=IFCCARTESIANPOINT((10.1,50.1));
#71=IFCCARTESIANPOINT((-20.2,60.2));
#72=IFCCARTESIANPOINT((-30.3,-70.3));
#73=IFCCARTESIANPOINT((40.4,-80.4));
#74=IFCDIRECTION((0.0,0.0,1.0));
#75=IFCRELCONTAINEDINSPATIALSTRUCTURE('ef8b3a5fde0a4f9fa61075474baa9568',#1,'ContainedInSpatial0',$,(#55),#38);
ENDSEC;
END-ISO-10303-21;
```

**Comment by aothms** [comment]
> As soon as a mandatory parameter is not found, an exception is thrown and IFCConvert terminates.

Yes, the structure of the code using exceptions has upsides and downsides. But we have try-catch blocks at various levels in the code. Somehow this one seeps through and isn't caught. Especially in the iterator https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcgeom/IfcGeomIteratorImplementation.h So for now I think the solution should be along those lines. Any help appreciated.

The long term solution is obviously drastic, where somehow we should only operate on valid subgraphs from the model. But the question is what constitutes a valid subgraph (if an irrelevant attribute is wrongly populated, can we still use that subgraph?) and what delineates the various subgraphs (if the material association is invalid, can we still process the geometry of the element for ex.?)

**Reply by aothms** [reply]
Edit:

> and IFCConvert terminates.

I just ran the code and this is apparently no the case though. In your sample model because there is only a single element, no output file is generated, but it's not an uncaught exception in the sense that the program immediately terminates. The exception is caught and the log messages are still printed.

```
Log:
[Notice] [2023-06-05 09:35:10] exclude entities IfcOpeningElement IfcSpace
[Warning] [2023-06-05 09:35:10] No plane angle unit encountered
[Warning] [2023-06-05 09:35:10] No representations encountered in relevant contexts, using all
[Error] [2023-06-05 09:35:10] {478ba5c0464b4fb9b63179c30dc05b27} Token $ at offset 2727 invalid instance name
[Notice] [2023-06-05 09:35:10] {478ba5c0464b4fb9b63179c30dc05b27} No geometrical elements found or none succesfully converted
```

**Reply by chetan-set** [reply]
Hi Thomas,

Thank you for your reply.

Even with a single element, an output `glb` file is created if we add the two fixes I mentioned in the original post.

Because that one `IFCBeam` is supposed to have a `IFCMaterial` (which is not present in the file), the `IFCBeam` is not generated and no `glb` file is generated (Hence the 'No geometrical elements found' notice).

I have done a test and commented out line 535 in `IFCParse.cpp`:

`throw IfcInvalidTokenException(t.startPos, toString(t), "instance name");`

followed by this `if` condition at line 558 of `IFCGeom.cpp`:
```
if (associated_material) {
...
}
```
Then the `IFCBeam` is converted to a `glb` file without its associated material and can be viewed and rendered just fine.

What I am hoping the ideal scenario would be is to have the `IFCBeam` rendered without a material.

What do you think?

**Comment by chetan-set** [comment]
Hi @aothms,

I have a small fix for this issue and would like to create a pull request.

I created a branch and I am trying to push it but I get this error:

```
%> git push origin handle_files_with_no_material 
remote: Permission to IfcOpenShell/IfcOpenshell.git denied to chetan-smart.
fatal: unable to access 'https://github.com/IfcOpenShell/IfcOpenshell.git/': The requested URL returned error: 403
```

Could I be given the permission to push a branch?

**Reply by Moult** [reply]
@chetan-set usually you fork the repository and push to your fork, then the PR can be merged in from your fork.

**Reply by aothms** [reply]
So after pushing the [fork] button in the github web ui you do `git remote add own https://github.com/chetan-set/IfcOpenShell` and `git push own`.

**Reply by chetan-set** [reply]
Thank you @Moult and @aothms. I managed to create my pull request and it can be found [here](https://github.com/IfcOpenShell/IfcOpenShell/pull/3270). Please give me your feedback
